### ISSUE TITLE
[TEST][S1] no KL / PPL-drift / tool-conformance gate, and the NIAH target has no invocation

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1571]
## Fact
- imp has no correctness gate that compares its output against a reference implementation: no KL divergence vs an fp16/bf16 HF reference, no perplexity-drift baseline, no CI-wired NIAH, no tool-schema conformance rate. Every instrument that exists is manual and none is referenced by any workflow.
- `.github/workflows/ci.yml:507`
- `scripts/validate_safetensors.py:11-14`
- `tests/refs/gen_reference.py:17-18`

## Evidence
- No KL divergence anywhere. The single occurrence of the token in the tree is in a "NOT executed" list:
```
$ grep -rniE "kl_div|kl-div|kl divergence|kldiv|kullback" . --exclude-dir=.git --exclude-dir=__pycache__ --exclude-dir=build --exclude-dir=build-dev
(no output)

$ sed -n '11,14p' scripts/validate_safetensors.py
NOT executed (Mode A):
  Phase 1 — BF16 reference (no BF16 checkpoints on disk)
  Phase 2 — NVFP4 calibration (imp consumes pre-quantized weights)
  Phase 5c — KL/PPL drift vs BF16 (no reference)

$ sed -n '860,861p' scripts/validate_safetensors.py
        "phase5c_drift_status": "INCOMPLETE — no BF16 reference (Mode A)",
        "phase5c_ppl_status": "INCOMPLETE — imp-server has no logprobs-of-arbitrary-text endpoint",
```
- No PPL baseline and no PPL threshold:
```
$ grep -rniE "ppl_baseline|perplexity_baseline|baseline_ppl|ppl_threshold|ppl_regress|max_ppl|ppl_delta" . --exclude-dir=.git --exclude-dir=build --exclude-dir=build-dev
(no output)

$ grep -rni "perplexity" Makefile scripts/ .github/
(no output)

$ grep -rn "ppl" tests/perf_baseline*.json
(no output)
```
  The only committed PPL tolerances compare imp to imp at a different chunk size, not to a reference:
```
$ sed -n '155,163p' tests/test_chunked_prefill.cu
// Chunk-invariance tolerances on teacher-forced PPL, relative to chunk=0.
// Measured cross-kernel-path noise (2026-06-06, Qwen3-8B + Llama-3.2-3B,
// 3.3-3.4k-token forcing): |dPPL| <= 0.15% relative. A positional or
// KV-corruption bug in the chunked path shifts PPL by >= 50%. 1% keeps a
// 6x noise margin while catching real bugs with > 50x margin.
static constexpr double kFp16RelTol = 0.01;
// FP8-KV adds dequant noise on the chunk-continuation KV reads.
static constexpr double kFp8RelTol = 0.03;
```
  `tools/analysis/awq_ppl_ab.sh:36-40` runs `imp-cli --perplexity` twice and greps the number out — it has no comparison, no arithmetic and no `exit 1` on a delta.
- The HF-reference generator that does exist is dormant by its own header:
```
$ sed -n '17,18p' tests/refs/gen_reference.py
STATUS: dormant infrastructure — no test currently
consumes these outputs; kept for future HF layer-by-layer comparisons.
```
- The two layer-diff tools compare imp against imp or against a llama.cpp text log, and neither exits non-zero:
  - `tools/analysis/layer_ab_diff.py:3` — "Per-layer, per-snapshot divergence between two runs of the SAME architecture"; `:151` returns 0 on the success path, `:134` only writes a marker.
  - `tools/analysis/layer_diff.py:3-6` — imp `.npy` dumps vs `llama-eval-callback` stderr; `main()` at `:245-282` has no return, so the process always exits 0.
- NIAH exists in two forms; neither is a CI gate:
```
$ sed -n '259,265p' tools/eval/niah/niah_bench.py
    (out_dir / "niah_results.json").write_text(
        json.dumps([dataclasses.asdict(r) for r in results], indent=2),
        encoding="utf-8",
    )
    write_summary(results, out_dir / "niah_summary.md")
    print(f"Wrote {out_dir}/niah_results.json and niah_summary.md", file=sys.stderr)
    return 0

$ sed -n '90,94p' tools/analysis/niah_check.py
    return 1 if fails else 0


if __name__ == "__main__":
    sys.exit(main())

$ grep -rn "niah_bench" Makefile scripts/ .github/ tests/ CMakeLists.txt
(no output)

$ grep -rni "niah\|degen_suite\|validate_safetensors\|awq_ppl\|perplexity\|kl_div" .github/
(no output)
```
  `niah_check.py` IS blocking under `make test-niah` (`Makefile:305-314`, last command of the recipe, `.SHELLFLAGS := -o pipefail -c` at `Makefile:7`) — but `test-niah` is a prerequisite of nothing:
```
$ grep -rn "test-niah" scripts/ .github/
(no output)
```
  `niah_check.py`'s own threshold is the strictest possible — 0 failed probes, i.e. 100% recall — and that part is a real gate, delivered by #1022. The defect is that nothing invokes it. The separate reporting tool `niah_bench.py` has no threshold at all, and its committed sample result records 60% at the long end with no verdict attached:
```
$ sed -n '14,17p' tools/eval/niah/sample_results/niah_summary.md
| fp16 | 100.0% | 60.0% |
| fp8 | 100.0% | 60.0% |
| nvfp4 | 100.0% | 66.7% |
| mxfp4_kv | 100.0% | 60.0% |
```
- No tool-schema conformance RATE anywhere. `degen_suite.py` has a `constrained` category with per-check booleans (`tools/analysis/degen_suite.py:700-708`, `:735-736`), no rate and no threshold, and `json_mode` is not a category:
```
$ grep -rn "json_mode" tools/analysis/degen_suite.py
(no output)
```
- The lane that would run the GPU-side asserts is off by default:
```
$ sed -n '505,510p' .github/workflows/ci.yml
    # skipped rather than failing. Until then,
    # this job is skipped — same effective behavior as the prior `if: false`,
    # but flips automatically when a runner appears. Also skipped on docs-only
    # changes (build uploads no artifact then — nothing to download/test).
    if: ${{ vars.HAS_GPU_RUNNER == 'true' && needs.build.outputs.code == 'true' }}
    runs-on: [self-hosted, gpu, cuda]
```
  and the repo already records this as an owner decision, with the consequence spelled out:
```
$ grep -n "F-5 (rest)" docs/audit/SETTLED.md
- **F-5 (rest)** — GPU CI lane: **declined by the repo owner, 2026-08-03.** The job and its
  nightly trigger stay dormant in `.github/workflows/ci.yml`. Consequence: `make verify-fast`
  locally is the only thing that ever runs a CUDA kernel against correctness or perf.
```
- `make test-golden` is a stub that asserts nothing and exits 0:
```
$ sed -n '327,329p' Makefile
# Golden output comparison (greedy, temp=0)
test-golden: build
	@echo "Golden output tests require running server — use pytest tests/api/ instead"
```

## Impact
- Per-arch/per-quant numerical correctness is unguarded. The repo's own history is what that costs: #1287 ("imp's SafeTensors path is 41% off on a GDN hybrid — proven against an HF reference"), #1273 ("Hybrid (GDN) + NVFP4 degrades badly on every checkpoint tested, both quantizers"), #1282 ("loads, reports tying success, then emits uniform logits, PPL == vocab_size"), #981 (a flag combination moving PPL 8.30 → 15.35). Every one of them was found by a human noticing bad output, not by a gate.
- `docs/MODELS.md` carries per-model quality claims (Gemma-4 NVFP4 "~+48% prose PPL vs UD-Q4_K_M", DeepSeek-V2-Lite "within ~3% of HF bf16", Qwen3.8-27B "PPL 4.62 on ppl_corpus_45k") that nothing re-measures. Those numbers can only go stale silently.
- Long-context retrieval is the ONE category with a real threshold (`niah_check.py`, 0 failed probes, from #1022) and it is still reachable only by a human typing `make test-niah`. The reporting tool beside it (`niah_bench.py`) records 60% recall at the long end with no verdict, which is what a category looks like when the instrument outlives the decision.
- Blast radius: all 16 architectures, all quant formats (GGUF Q*_K/Q8_0/MXFP4, NVFP4 Modelopt + llm-compressor, FP8 sources), every endpoint.

## Repro
```bash
# NOT RUN — all of these need a GPU and a checkpoint; scheduling is the caller's.
make test-niah NIAH_MODEL=Qwen3-14B-Q6_K.gguf     # exercises the one real gate that exists
python3 tools/analysis/degen_suite.py --url http://localhost:8080
bash tools/analysis/awq_ppl_ab.sh                  # prints two PPLs, decides nothing
```

## Fix direction
- Add one reference-parity gate, not five. The cheapest with a real oracle: teacher-forced NLL of `imp-cli --perplexity` on `tools/analysis/ppl_corpus_45k.txt` against a committed per-(arch, quant) baseline, with a threshold set from measured noise the way `tests/test_chunked_prefill.cu:155-162` already sets its 1%/3%. `imp_perplexity` is already the project's documented determinism-proof instrument (`docs/determinism.md`, "Shipped guarantees").
- NIAH does not need a threshold — it has one. `make test-niah` runs `niah_check.py`, which exits 1 on any failed probe (`tools/analysis/niah_check.py:90`). It needs an *invocation*: a CI job, a nightly dispatch, or a staleness check in `scripts/check-release.sh`. Do not confuse it with `tools/eval/niah/niah_bench.py`, which reports and returns 0 unconditionally (`:265`).
- The GPU CI lane being declined (SETTLED F-5) does not force this to be nothing: a nightly `workflow_dispatch` on the existing self-hosted job definition, or a committed result file with a date that `scripts/check-release.sh` fails on when stale, both put a floor under it without a permanent runner.

## Acceptance criteria
- [ ] At least one gate compares imp against a non-imp reference and fails the build on a threshold. Name the file, the threshold number and the command.
- [ ] `grep -rn "<gate script>" .github/workflows/` returns a non-empty result, or the equivalent staleness gate exists in `scripts/check-release.sh` with the same effect.
- [ ] `make test-niah` is reachable from an automatic trigger; `grep -rn "test-niah" .github/ scripts/` returns a non-empty result, or `docs/` states explicitly that it is operator-run and `scripts/check-release.sh` fails when the last recorded run is stale.
- [ ] Mutation-validated: reintroducing the #1282 defect class (tied-embedding load that yields uniform logits) makes the new gate red. Paste the failing output.

## Unverified
- Whether `vars.HAS_GPU_RUNNER` is set on the GitHub repo cannot be read from the checkout. `docs/audit/SETTLED.md` §G records the lane as declined by the owner on 2026-08-03 and dormant, which is the in-tree evidence; a `gh api repos/kekzl/imp/actions/variables` call would settle it.

**Comment by kekzl** [comment]
Verified the two structural sub-claims on CPU; both hold.

**`test-niah` is defined and invoked by nothing.** `Makefile:315` (`test-niah: build check-gpu`), and no workflow, script or doc calls it:

```
$ grep -rn 'test-niah' .github/workflows/ scripts/ docs/
(the only hits are docs/audit/… and docs/archive/… referring to the README's staleness, not to running it)
```

**No perplexity wiring at all**, same shape:

```
$ grep -rn 'perplexity\|--ppl' Makefile .github/workflows/*.yml
(no output)
```

So the target is not "loosely gated", it is a target nobody runs — the same class as #1299 (the determinism gate that never executed) and the `-m nomodel` marks in #1600. The instruments exist; nothing invokes them.

The gate itself still needs what this issue says it needs: a GPU runner and a reference model for KL/PPL. That part is not CPU-closable.

**Comment by kekzl** [comment]
Closing as won't-fix: the instrument cannot be built in CI, and the gap is now
recorded as a limitation instead of as work in flight.

Every gate this asks for needs a card:

- KL divergence and PPL drift need a BF16 reference forward.
  `scripts/validate_safetensors.py:11-14` already names why it cannot run one:
  no BF16 checkpoint is on disk, and imp consumes pre-quantised weights.
- `make test-niah` exists (`Makefile:315`). Nothing invokes it because, like
  every target carrying `check-gpu` among its prerequisites, it boots a model.
  That is the same wall, not a separate oversight: wiring it into a workflow
  would produce a job that cannot run.

CI has no GPU runner and none is planned. `make verify-fast` on a machine with
a card is the only lane in this repo that runs a kernel at all.

Recorded in `docs/LIMITATIONS.md` under "Gates that do not exist" (PR below),
so a reader deciding against imp sees it without reading the issue tracker.

Reopen if a GPU runner or a BF16 reference lands.

PR: #1736
